tools/halide_image_io.h

```diff
@@ -959,8 +959,150 @@ bool find_imageio(const std::string &filename, ImageIO<ImageType, check> *result
     return check(false, err.c_str());
 }
 
+// Given something like ImageType<Foo>, produce typedef ImageType<Bar>
+template<typename ImageType, typename ElemType>
+struct ImageTypeWithElemType {
+    using type = decltype(std::declval<ImageType>().template as<ElemType>());
+};
+
+// Must be constexpr to allow use in case clauses.
+inline constexpr int halide_type_code(halide_type_code_t code, int bits) {
+    return (((int) code) << 8) | bits;
+}
+
 }  // namespace Internal
 
+struct ImageTypeConversion {
+    // Convert an Image from one ElemType to another, where the src and
```

Comment:
This could be templatized even further to support that (move the convert<> methods into a template class); I haven't seen a need for such a conversion yet, so I haven't written one.